server/sonar-server/src/main/java/org/sonar/server/issue/ws/IssueDataLoader.java

```diff
@@ -0,0 +1,177 @@
+import org.sonar.db.user.UserDto;
+import org.sonar.server.user.UserSession;
+
+/**
+ * Loads issues and related information from database.
+ */
+public class IssueDataLoader {
+
+  private final DbClient dbClient;
+  private final UserSession userSession;
+
+  public IssueDataLoader(DbClient dbClient, UserSession userSession) {
+    this.dbClient = dbClient;
+    this.userSession = userSession;
+  }
+
+  public Data load(List<String> issueKeys, Selection selection) {
+    Data result = new Data();
+    DbSession dbSession = dbClient.openSession(false);
+    try {
+      result.issues = dbClient.issueDao().selectByOrderedKeys(dbSession, issueKeys);
+      Collector collector = collect(result.issues, selection);
+      if (selection.contains(SelectionKey.RULE)) {
+        result.rules = dbClient.ruleDao().selectByKeys(dbSession, collector.<RuleKey>get(SelectionKey.RULE));
+      }
+      if (selection.contains(SelectionKey.COMMENT)) {
+        result.comments = dbClient.issueChangeDao().selectByTypeAndIssueKeys(dbSession, issueKeys, IssueChangeDto.TYPE_COMMENT);
+        for (IssueChangeDto comment : result.comments) {
+          collector.add(SelectionKey.USER, comment.getUserLogin());
+        }
+      }
+      if (selection.contains(SelectionKey.USER)) {
+        result.users = dbClient.userDao().selectByLogins(dbSession, collector.<String>get(SelectionKey.USER));
+      }
+      if (selection.contains(SelectionKey.ACTION_PLAN)) {
+        result.actionPlans = dbClient.actionPlanDao().selectByKeys(dbSession, collector.<String>get(SelectionKey.ACTION_PLAN));
+      }
+      if (selection.contains(SelectionKey.COMPONENT) || selection.contains(SelectionKey.PROJECT)) {
+        result.components = dbClient.componentDao().selectByKeys(dbSession, collector.<String>get(SelectionKey.COMPONENT));
+        result.components.addAll(dbClient.componentDao().selectSubProjectsByComponentUuids(dbSession, collector.<String>get(SelectionKey.COMPONENT)));
+        for (ComponentDto component : result.components) {
+          collector.add(SelectionKey.PROJECT, component.projectUuid());
+        }
+        if (selection.contains(SelectionKey.PROJECT)) {
+          result.projects = dbClient.componentDao().selectByKeys(dbSession, collector.<String>get(SelectionKey.PROJECT));
+          result.components.addAll(result.projects);
+        }
+      }
+      return result;
+    } finally {
+      dbClient.closeSession(dbSession);
+    }
+  }
+
+  private Collector collect(List<IssueDto> issues, Selection selection) {
+    Collector collector = new Collector();
+    collector.collect(issues);
+    collector.collect(selection);
+    if (userSession.isLoggedIn()) {
+      collector.add(SelectionKey.USER, userSession.getLogin());
+    }
+    return collector;
+  }
+
+  public static class Data {
+    private List<IssueDto> issues;
+    private List<ComponentDto> components;
+    private List<ComponentDto> projects;
+    private List<UserDto> users;
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Remove this unused "users" private field. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1068)